### PR TITLE
mirage: 0.5.2 -> 0.6.4

### PR DESCRIPTION
**Pull request by vojta001**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vojta001** [comment]
@colemickens

**Comment by vojta001** [comment]
The only problematic part is QML cache. Mirage stores its QML caches in `.cache/mirage/mirage/qmlcache`, however it is not purged on update, causing problems. I suspect it is related to the fact the mirage binary in the Nix store has the creation timestamp 0 (as all files in the store). Can we do anything about it?

**Comment by symphorien** [comment]
Works after clearing the cache

**Comment by colemickens** [comment]
We could write a wrapper that writes "last-booted-version" to a file in cache near (in?) mirage's dir. If we see it's different on next boot, we wipe the mirage cache dir before starting?

**Comment by symphorien** [comment]
Or maybe we can patch the path to the cache to be `.cache/mirage/$out` ? Is there a XDG_SOMETHING_DIR variable that we can set in the wrapper ?

**Comment by mirukana** [comment]
There are some environment variable to disable or relocate the cache: https://doc.qt.io/qt-5/qmldiskcache.html

**Comment by symphorien** [comment]
Thanks !
This diff solves the segfault:
```diff
diff --git a/pkgs/applications/networking/instant-messengers/mirage/default.nix b/pkgs/applications/networking/instant-messengers/mirage/default.nix
index de84f3a0405..c204ca720e3 100644
--- a/pkgs/applications/networking/instant-messengers/mirage/default.nix
+++ b/pkgs/applications/networking/instant-messengers/mirage/default.nix
@@ -45,6 +45,7 @@ mkDerivation rec {
     buildPythonPath "$out $pythonPath"
     wrapProgram $out/bin/mirage \
       --prefix PYTHONPATH : "$PYTHONPATH" \
+      --run 'export QML_DISK_CACHE_PATH=''${XDG_CACHE_HOME:-$HOME/.cache}/mirage/qmlcache/${placeholder "out"}' \
       "''${qtWrapperArgs[@]}"
     '';
 ```

**Comment by mirukana** [comment]
Try also https://github.com/mirukana/mirage/issues/90#issuecomment-727023274, if this causes no issues we can use the flag by default for future releases.

**Comment by symphorien** [comment]
Indeed with 
```diff
diff --git a/pkgs/applications/networking/instant-messengers/mirage/default.nix b/pkgs/applications/networking/instant-messengers/mirage/default.nix
index de84f3a0405..c57e57d5d75 100644
--- a/pkgs/applications/networking/instant-messengers/mirage/default.nix
+++ b/pkgs/applications/networking/instant-messengers/mirage/default.nix
@@ -38,7 +38,7 @@ mkDerivation rec {
 
   pythonPath = pypkgs;
 
-  qmakeFlags = [ "PREFIX=${placeholder "out"}" ];
+  qmakeFlags = [ "PREFIX=${placeholder "out"}" "CONFIG+=qtquickcompiler" ];
 
   dontWrapQtApps = true;
   postInstall = ''
```
I can alternate between stable's mirage and this PR without segfault.

**Comment by SuperSandro2000** [comment]
@vojta001 Can you update your PR with the cache fixes?

**Comment by vojta001** [comment]
@SuperSandro2000  And what solution do we prefer? Is it disabling the cache completely (probably slower, but does not fill the home dir with mess), or adding the derivation hash to the cache name?

**Comment by symphorien** [comment]
My preference goes to https://github.com/NixOS/nixpkgs/pull/99121#issuecomment-727165392

**Comment by vojta001** [comment]
Here you go. Please review @colemickens

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 99121` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mirage-im</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
on darwin:

```
clarative-5.12.7/lib/libQt5Qml.dylib /nix/store/qafg270zx0ih7rr18gxfvx0b8jb2dnfw-qtbase-5.12.7/lib/libQt5Network.dylib /nix/store/qafg270zx0ih7rr18gxfvx0b8jb2dnfw-qtbase-5.12.7/lib/libQt5Core.dylib -framework OpenGL -framework AGL
@nix { "action": "setPhase", "phase": "glibPreInstallPhase" }
glibPreInstallPhase
@nix { "action": "setPhase", "phase": "installPhase" }
installing
install flags: SHELL=/nix/store/k89nm2jva0qmvd970f84wq2iq1iwm9bs-bash-4.4-p23/bin/bash gsettingsschemadir=/nix/store/dkrwri1an9582pmpbz8c4250qdkzcmgc-mirage-0.5.2/share/gsettings-schemas/mirage-0.5.2/glib-2.0/schemas/ install
Makefile:835: warning: overriding recipe for target 'install_icons256'
Makefile:817: warning: ignoring old recipe for target 'install_icons256'
Makefile:839: warning: overriding recipe for target 'uninstall_icons256'
Makefile:821: warning: ignoring old recipe for target 'uninstall_icons256'
/nix/store/gzxz5xbx5bcvy5lvh8g9wv0871ipi9na-qtbase-5.12.7-dev/bin/qmake -install qinstall /private/tmp/nix-build-mirage-0.5.2.drv-0/source/extra/linux/mirage.png /nix/store/dkrwri1an9582pmpbz8c4250qdkzcmgc-mirage-0.5.2/share/icons/hicolor/256x256/apps/mirage.png
/nix/store/gzxz5xbx5bcvy5lvh8g9wv0871ipi9na-qtbase-5.12.7-dev/bin/qmake -install qinstall /private/tmp/nix-build-mirage-0.5.2.drv-0/source/extra/linux/mirage.desktop /nix/store/dkrwri1an9582pmpbz8c4250qdkzcmgc-mirage-0.5.2/share/applications/mirage.desktop

Builder called die: Cannot wrap '/nix/store/dkrwri1an9582pmpbz8c4250qdkzcmgc-mirage-0.5.2/bin/mirage' because it is not an executable file
Backtrace:
7 assertExecutable /nix/store/g5nqhchcnq9c9ym8ihqj8v5mbbibsa3x-hook/nix-support/setup-hook
136 wrapProgram /nix/store/g5nqhchcnq9c9ym8ihqj8v5mbbibsa3x-hook/nix-support/setup-hook
90 _callImplicitHook /nix/store/m40i7gdarksan1ax23vvwrq17p3pxjb3-stdenv-darwin/setup
106 _eval /nix/store/m40i7gdarksan1ax23vvwrq17p3pxjb3-stdenv-darwin/setup
49 runHook /nix/store/m40i7gdarksan1ax23vvwrq17p3pxjb3-stdenv-darwin/setup
1118 installPhase /nix/store/m40i7gdarksan1ax23vvwrq17p3pxjb3-stdenv-darwin/setup
1309 genericBuild /nix/store/m40i7gdarksan1ax23vvwrq17p3pxjb3-stdenv-darwin/setup
2 main /nix/store/9krlzvny65gdc8s7kpb6lkx8cd02c25b-default-builder.sh
```

**Comment by vojta001** [comment]
@SuperSandro2000 What do you suggest? Adding the `qtquickcompiler` for Linux only?

**Comment by symphorien** [comment]
Mirage has never built on darwin so let's remove darwin from meta.platforms.

https://hydra.nixos.org/job/nixpkgs/trunk/mirage-im.x86_64-darwin/all

**Comment by SuperSandro2000** [comment]
Either mark it broken if upstream has darwin support. If upstream does not mention support for darwin remove it from platforms.

**Comment by vojta001** [comment]
@SuperSandro2000 Done. Their support of Darwin isn't the best, but they claim it to be doable and there are unofficial guides for building on Darwin

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 99121` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mirage-im</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 99121` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>mirage-im</li>
  </ul>
</details>

File: pkgs/applications/networking/instant-messengers/mirage/default.nix [hunk 1]
```diff
@@ -1,7 +1,8 @@
-{ lib, mkDerivation, fetchFromGitHub
+{ lib, mkDerivation, fetchFromGitHub, stdenv
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/mirage/default.nix:1]:
```suggestion
{ stdenv, mkDerivation, fetchFromGitHub
```

Also update meta with to stdenv.lib.

File: pkgs/applications/networking/instant-messengers/mirage/default.nix [hunk 2]
```diff
@@ -1,7 +1,8 @@
...
 , qmake, pkgconfig, olm, wrapQtAppsHook
 , qtbase, qtquickcontrols2, qtkeychain, qtmultimedia, qttools, qtgraphicaleffects
 , python3Packages, pyotherside
+, libXScrnSaver
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/mirage/default.nix:5]:
```suggestion
, python3Packages, pyotherside, libXScrnSaver
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by symphorien** [comment]
can you solve the merge conflict ?

**Comment by vojta001** [comment]
Shall I merge master into my feature branch or shall I rebase onto master?

**Comment by symphorien** [comment]
please rebase. You may find it easier if you squash your commits before rebasing.

**Comment by vojta001** [comment]
Done

**Comment by vojta001** [comment]
Could you please test the resulting binary? I have problems with it now and I don't know whether they are related to the state of my `.cache` folder or whether the build is broken.

**Comment by symphorien** [comment]
works for me. How does it fail on your end?

**Comment by vojta001** [comment]
The windows launches (therefore QML cache works just fine) and displays

`Unexpected error occured: OlmAccountError`

```
load_saved_accounts.<some uid>

Traceback (most recent call last):
  File "qrc:/src/backend/qml_bridge.py", line 83, in on_done
    result = future.result()
  File "/nix/store/cqh3nsk4z8lc960g8yqvz5nf2f6qx9v0-python3-3.8.6/lib/python3.8/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/nix/store/cqh3nsk4z8lc960g8yqvz5nf2f6qx9v0-python3-3.8.6/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
  File "qrc:/src/backend/backend.py", line 295, in load_saved_accounts
    return await asyncio.gather(*(
  File "qrc:/src/backend/backend.py", line 284, in resume
    await self.resume_client(
  File "qrc:/src/backend/backend.py", line 272, in resume_client
    await client.resume(user_id, token, device_id, state, status_msg)
  File "qrc:/src/backend/matrix_client.py", line 311, in resume
    self.restore_login(user_id, device_id, access_token)
  File "/nix/store/qssk7fsipsl6chwfr3kn9fcx9yp0pr7d-python3.8-matrix-nio-0.15.1/lib/python3.8/site-packages/nio/client/base_client.py", line 406, in restore_login
    self.load_store()
  File "/nix/store/qssk7fsipsl6chwfr3kn9fcx9yp0pr7d-python3.8-matrix-nio-0.15.1/lib/python3.8/site-packages/nio/client/base_client.py", line 382, in load_store
    self.olm = Olm(self.user_id, self.device_id, self.store)
  File "/nix/store/qssk7fsipsl6chwfr3kn9fcx9yp0pr7d-python3.8-matrix-nio-0.15.1/lib/python3.8/site-packages/nio/crypto/olm_machine.py", line 213, in __init__
    account = self.store.load_account()  # type: ignore
  File "/nix/store/qssk7fsipsl6chwfr3kn9fcx9yp0pr7d-python3.8-matrix-nio-0.15.1/lib/python3.8/site-packages/nio/store/database.py", line 166, in load_account
    return OlmAccount.from_pickle(
  File "/nix/store/qssk7fsipsl6chwfr3kn9fcx9yp0pr7d-python3.8-matrix-nio-0.15.1/lib/python3.8/site-packages/nio/crypto/sessions.py", line 45, in from_pickle
    account = super().from_pickle(pickle, passphrase)
  File "/nix/store/bxipga1mp2s3vgclma5linvjyn0ailwf-python3.8-python-olm-3.1.5/lib/python3.8/site-packages/olm/account.py", line 166, in from_pickle
    obj._check_error(ret)
  File "/nix/store/bxipga1mp2s3vgclma5linvjyn0ailwf-python3.8-python-olm-3.1.5/lib/python3.8/site-packages/olm/account.py", line 99, in _check_error
    raise OlmAccountError(last_error)
olm.account.OlmAccountError: UNKNOWN_PICKLE_VERSION
```

**Comment by symphorien** [comment]
Wild guess: did you run a version of mirage using a later version of python which has introduced a new version of the pickle format?
@mirukana any idea of the cause?

**Comment by vojta001** [comment]
It is possible. I have compiled and run multiple variants during the development of this PR. Interestingly the current v0.5.2 works fine for me.

If the problem is on my machine only, let's just proceed and I'll simply wipe my `.cache/mirage` folder.

**Comment by symphorien** [comment]
@GrahamcOfBorg build mirage-im

**Pull request merged**